### PR TITLE
Don't allow 64bit-indices in PreconditionAMGMueLu

### LINKED ISSUE #3968: Test failure: linear_operator_10
This fails when the library is compiled with 64 bit indices with the following message:

```
5065:
5065: terminate called after throwing an instance of 'std::runtime_error'
5065:   what():  /home/drwells/Documents/Code/CPP/trilinos/packages/xpetra/src/Map/Xpetra_EpetraMap.hpp:678:
5065:
5065: Throw number = 1
5065:
5065: Throw test that evaluated to true: !map->GlobalIndicesIsType<GlobalOrdinal>()
5065:
5065: Xpetra::EpetraMapT: GlobalOrdinal mismatch.
```

I believe that the issue is in the MueLu wrappers: we don't seem to create matrices with 64 bit indices in `PreconditionAMGMueLu::initialize`. This should not be hard to fix but I don't know Xpetra well enough to do it myself.

Ping @Rombur

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #3968.

### CODE REVIEW

File: include/deal.II/lac/trilinos_precondition.h [hunk 1]
```diff
@@ -1618,7 +1618,8 @@ namespace TrilinosWrappers
    * except for the higher_order_elements parameter which does not exist in
    * PreconditionerAMGMueLu.
    *
-   * This class requires Trilinos 11.14 or higher.
+   * This class requires Trilinos 11.14 or higher. At the moment 64bit-indices
```
Comment (jppelteret) [inline include/deal.II/lac/trilinos_precondition.h]: Should we make this more visible with a `@note`, and the warning about the unstable interface with a `@warning`?

**Comment by jppelteret** [comment]
I can't comment on it directly, but you've renamed the test output files with the suffix `with_64bit_indices_off` as opposed to `with_64bit_indices=off`, which is how I see other tests in `test/trilinos` named. Does this matter?

**Comment by Rombur** [comment]
Can you add the same warning for ML.

**Comment by masterleinad** [comment]
@jppelteret Yes, it matters. Seems this got lost while cherry-picking. Thanks!

**Comment by masterleinad** [comment]
@Rombur You mean `TrilinosWrappers::PreconditionAMG` doesn't work with 64bit-indices? At least for me `trilinos/precondition_amg_smoother` also works with 64bit-indices.

**Comment by Rombur** [comment]
When I wrote the wrapper for MueLu, I didn't support 64 bit indices because ML didn't. As far as I can tell this hasn't changed.

**Comment by Rombur** [comment]
Hmm actually, I looked in my old notes and at the time I wrote the wrappers Xpetra didn't support Epetra64. Later today, I have a meeting with one of the developer of ML/MueLu. I will ask him about 64bit support in ML.

**Comment by Rombur** [comment]
I just asked him and ML does not support Epetra64.

**Comment by masterleinad** [comment]
Hmm... it seems that `amg_nondeterministic_02` is already using `TrilinosWrappers::PreconditionAMG` and this test is not restricted to 32bit indices. Furthermore, the tests `trilinos/precondition_amg_smoother` and `trilinos/precondition_amg_dgp` are working also for `64bit-indices`. Hence, I am a bit hesitant to use the same restriction in `TrilinosWrappers::PreconditionAMG`.

Do you have an explanation for this?

**Comment by bangerth** [comment]
Can we leave this for a separate PR? I think the current patch as is is strictly progress, and I would like to merge it.

I don't recall but it seems to me that we've done some runs with ASPECT and 64-bit. In that case, we just use the Epetra64 functions and ML will have to determine what it wants to do with this. We could try out what exactly happens in that case, but that is an issue for a separate PR.

**Comment by Rombur** [comment]
If you don't believe me, just change the test to solve a problem with more than two billions unknowns and see if it works... The 64bit tests never check that the code works with more than two/four billions of unknowns. If somewhere a `unsigned long long int` is casted to a `unsigned int`, you will only see the problem when the number of dofs is greater than 4 billions. We just give a `Epetra_CrsMatrix` and a `Epetra_MultiVector` to `ML`. I don't know what they are doing internally.

**Comment by bangerth** [comment]
I trust you alright :-) If you say it doesn't work, it almost certainly doesn't. And you're probably right that we simply never tried it with that many elements.

My statement was more than if (i) we compile with 64-bit indices, (ii) still only use a small number of DoFs, then (iii) ML seemed to work as far as I recall. If I understand you correctly, then the condition that makes ML fail is not using 64-bit indices, but actually *exceeding* the 32-bit index space?

**Comment by Rombur** [comment]
Yes. The reason it crashes for MueLu is that we use Xpetra which uses a template parameter for the GlobalOrdinal. For ML, we use Epetra. The difference between 32 and 64 bits supports is the name of some functions (ex. you can call GlobalLength() or GlobalLength64()). As far as I can tell nothing forbids you to call GlobalLength() on a vector with more than 2 billions of unknowns...

**Comment by luca-heltai** [comment]
I can confirm that @nicola-giuliani observed the same problem

**Pull request merged**